### PR TITLE
CDAP-1411 refactor reflection writer/reader classes so that they

### PR DESCRIPTION
**Pull request by albertshau**
can be re-used for formats other than our avro-like binary
encoding. Also adding classes that encode objects as Put objects
to be written to a Table, and decode objects from Row objects
fetched from a Table.

### CODE REVIEW

**Comment by albertshau** [comment]
http://builds.cask.co/browse/CDAP-DUT766

File: cdap-common/src/main/java/co/cask/cdap/internal/io/ReflectionDatumWriter.java [hunk 1]
```diff
@@ -55,204 +44,111 @@ public Schema getSchema() {
...
-      }
-      size = col.size();
-    } else {
-      size = Array.getLength(array);
-      encoder.writeInt(size);
-      for (int i = 0; i < size; i++) {
-        write(Array.get(array, i), encoder, componentSchema, seenRefs);
-      }
+  @Override
+  protected void writeArray(Encoder encoder, Collection col, Schema componentSchema) throws IOException {
```
Comment (chtyim) [inline cdap-common/src/main/java/co/cask/cdap/internal/io/ReflectionDatumWriter.java]: Always use generic type `Collection<?>`, even it's wildcard.

File: cdap-common/src/main/java/co/cask/cdap/internal/io/ReflectionReader.java [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+import java.util.Map;
+import java.util.UUID;
+
+/**
+ * Reflection based reader.
+ *
+ * @param <SOURCE> type of object to read from
+ * @param <TO> type of object to read
+ */
+public abstract class ReflectionReader<SOURCE, TO> {
```
Comment (chtyim) [inline cdap-common/src/main/java/co/cask/cdap/internal/io/ReflectionReader.java]: Maybe call it `FROM`, `TO`?

**Comment by chtyim** [comment]
Overall LGTM, just two minor comments.

**Pull request merged**